Minimap plugin

Danielv123 opened the pull request:
Adds a mapping plugin, drawing a representation of the mapview of instances in the web UI with support for timelapses.
It has been a priority to support history data while keeping the runtime performance of the instance and controller and storage consumption on the controller as low as possible, making some tradeoffs for web client rendering performance.

It heavily relies on custom binary formats to achieve the storage footprint desired.

Ideally it will be good enough to be left enabled by default without having to think about it to later be able to use it to review history for anti griefing purposes or making a neat timelapse of a multiplayer map.

One thing that is unique about the timelapsing feature as implemented here is that we only store changes, so a replay would skip over periods when players are taking a break without additional editing effort.

My intention is for this to be able to be used as a base for the web interface and other plugins. Some examples:

- The instance page could show a small window of the main factory. this would make it easier to visually tell instances apart
- Gridworlds should render multiple instances spread out on one canvas so it looks like one
- Universal edges should render its edges on the map as another overlay layer - not sure how to integrate that cleanly

Todo:
- [x] Tracking players
- [ ] More testing with multiple surfaces
- [ ] Testing performance with a longer history (ideally a full game playthrough)
- [x] Clean up user interface
- [x] Button to delete history, for example if you are switching savefile on an instance or the history gets too large
- [ ] More granular subscriptions - only send updates for clients viewing the right instance/surface/force

## Changelog
```
### Features
- Added minimap to web interface. [#999](https://github.com/clusterio/clusterio/issues/999)
```

Danielv123 committed:
Create minimap core plugin based on gridworld plugin

Danielv123 committed:
Use new chunk charting API

Danielv123 committed:
Cleanup, web interface now shows images (but corrupted)

Danielv123 committed:
Improve webUI loading speed with 512x512 tiles, still corrupted looking

Danielv123 committed:
Use encode_string and support multiple instances

Danielv123 committed:
Render map tiles as pixelated

Danielv123 committed:
Switch to custom canvas based map renderer

Danielv123 committed:
Fix negative coordinates when drawing updated chunks

Danielv123 committed:
Fix issues with black seams from incorrect scaling math

Danielv123 committed:
Allow click and drag to pan

Danielv123 committed:
Smooth keyboard movement

Danielv123 committed:
Add readme

Danielv123 committed:
Document new data flow and tile storage format

Danielv123 committed:
Implement new tile storage format with tentative timelapse support

Danielv123 committed:
Calculate and store deltas for changes

Danielv123 committed:
Use same parsing utilities on backend and frontend

Danielv123 committed:
Add timeline selection components and track tick on changes

Danielv123 committed:
Working timelapse rendering functionality

Danielv123 committed:
Incremental rendering of timeline scrubbing

Danielv123 committed:
F and G as timeline navigation shortcuts

Danielv123 committed:
Optimize canvas to render independent of react

Danielv123 committed:
Add chart tag support

Also fixes react state management issues, moving a few things over to useRef and showing all instances instead of using a custom solution

Danielv123 committed:
Fix deleting and modifying tags, tags on timeline

Danielv123 committed:
Implement recipe overlay

Danielv123 committed:
Fix recipe data deduplication

Danielv123 committed:
Timeline support for recipe overlay

Danielv123 committed:
Use binary storage for recipe overlay for 8x space reduction

Danielv123 committed:
Document tile data storage

Danielv123 committed:
Further compress tile pixel delta data

7 bytes header + 6 bytes per pixel -> 11 byte header + 2 bytes per pixel

Danielv123 committed:
Lint and refactor

Danielv123 committed:
Fix pixel change record bounds check

From cursor review: Insufficient bounds checking
for pixel change records (type 2) in renderTileToPixels,
extractAvailableTicks, and parseTileData. The code
checks for 6 bytes of header data but reads 10 bytes
(tick, pixelCount, newColor, oldColor), potentially
leading to reading out of bounds. The check should
be offset + 10 > tileData.length.

Danielv123 committed:
Fix pixel parsing offset error

Danielv123 commented:
For some reason previous_recipe.name.name is correct. Doesn't look like it makes sense but thats how it works

Danielv123 committed:
Add player position tracking

Danielv123 commented on plugins/minimap/controller.ts:
```diff
@@ -0,0 +1,1077 @@
+						buffer.writeUInt8((y_int24 >> 16) & 0xFF, offset);
+						buffer.writeUInt8((y_int24 >> 8) & 0xFF, offset + 1);
+						buffer.writeUInt8(y_int24 & 0xFF, offset + 2);
+						offset += 3;
```
No, its a 24 bit signed int as per the docs.

Danielv123 committed:
Clean up UI, fix focus issues

Danielv123 committed:
Attempt to fix tests

Danielv123 committed:
Don't load Lua in 1.1

Danielv123 committed:
Remove extraneous logging

Danielv123 committed:
Fix player ID assignments

Danielv123 committed:
Move player position parsing to shared util, fix player ID persistence

Danielv123 committed:
Remove unused filter

Danielv123 committed:
Fix lint, use of sync method in web UI

Danielv123 committed:
Use SubscriptionController

Danielv123 committed:
Fix incorrect type in player path request response

Danielv123 committed:
Use filtered subscriptions to save data

Danielv123 committed:
Add filtered subscriptions

Danielv123 committed:
Add tests for subscription filters

Danielv123 committed:
Merge remote-tracking branch 'origin/master' into plugin-minimap

Danielv123 committed:
Fix circular imort breaking web UI

Danielv123 committed:
Fix react errors in factorio version selector

Danielv123 committed:
Add button to delete minimap data

Danielv123 committed:
Add antd icons to minimap

Danielv123 committed:
Add antd icons to minimap

Danielv123 committed:
Merge branch 'plugin-minimap' of https://github.com/clusterio/clusterio into plugin-minimap

Danielv123 committed:
Stable label creation, forgot imports of messages

Danielv123 committed:
Don't break version selector

Danielv123 committed:
Player session end tracking

Danielv123 committed:
Extend minimap components to allow external use

Cooldude2606 commented on plugins/minimap/docs/tile_data_storage.md:
```diff
@@ -0,0 +1,118 @@
+| Bytes | Field               | Description                 |
+| ----- | ------------------- | --------------------------- |
+| 4     | `tick` (uint32)     | ⌊tick / 60⌋ timestamp       |
+| 2     | `count` (uint16)    | Number of pixels changed    |
```
I can't help but feel that this doesn't need to be 2 bytes. When can we expect more than a quater of a chunk (255 tiles) to be updated with the same transition? If needs be we could have a "Large" pixel change record as an extra type for some easy savings.

Danielv123 replied:
Good observation about the practical limits. A "Large" pixel change record variant could be a viable optimization if we encounter cases where >255 pixel transitions occur frequently. This would be worth revisiting in a follow-up if telemetry shows it is needed.

Cooldude2606 commented on plugins/minimap/module/minimap.lua:
```diff
@@ -0,0 +1,576 @@
+			elseif not last_position then
+				-- First position for this player
+				should_update = true
+			end
```
`last_position` is non-nil following the call to `has_moved_significantly`. No code error, just redudant checks and an unreachable block.

Cooldude2606 commented on plugins/minimap/module/minimap.lua:
```diff
@@ -0,0 +1,576 @@
+		storage.minimap.player_positions[player.index] = nil
+	end
+
+	-- The next position update will be sent when the player moves or in process_player_positions
```
How come there is no session start? If you add it on the first player pos update then you could be waiting upto 5 seconds before start session is recorded.

Cooldude2606 commented on plugins/minimap/module/minimap.lua:
```diff
@@ -0,0 +1,576 @@
+									return { type = rp.main_product.type, name = rp.main_product.name }
+								end
+								return nil
+							end)()
```
A closure like this isnt great when it would otherwise be two lines of code above `local data`

Cooldude2606 commented on plugins/minimap/package.json:
```diff
@@ -0,0 +1,57 @@
+{
+	"name": "@clusterio/plugin-minimap",
+	"version": "2.0.0-alpha.1",
```
Version must match current version on main, currently `22b`

Danielv123 replied:
✅ Confirmed: plugins/minimap/package.json version is `2.0.0-alpha.22b`, matching current main branch baseline.

Cooldude2606 commented on plugins/minimap/package.json:
```diff
@@ -0,0 +1,57 @@
+	},
+	"peerDependencies": {
+		"@clusterio/lib": "workspace:*"
+	},
```
Following the merge of #823 the peer deps here must include controller, webui, and host to have correct types. Similarly need to be added as dev deps, and existing deps checked for if `catalog:` should be used.

Cooldude2606 commented on plugins/minimap/index.ts:
```diff
@@ -0,0 +1,52 @@
+	title: "View Minimap",
+	description: "View the interactive minimap of Factorio instances.",
+	grantByDefault: true,
+});
```
Do you think we should split this into "view live" and "view historic" ?

Danielv123 committed:
Fix surfaces with underscore in the name

Cooldude2606 commented on plugins/minimap/web/dataSources/SingleInstanceDataSource.ts:
```diff
@@ -0,0 +1,191 @@
+			this.plugin.setInstanceSurfaceFilters(filters);
+		} else if (this.plugin.setInstanceSurfaceFilter) {
+			this.plugin.setInstanceSurfaceFilter(this.instanceId, this.surface);
+		}
```
Why do both of these exist?

Cooldude2606 commented on plugins/minimap/instance.ts:
```diff
@@ -0,0 +1,249 @@
+		if (output.type === "action" && output.action === "CUSTOM") {
+			// Could handle custom factorio output here if the Lua module sends special messages
+		}
+	}
```
Can be removed. Although what where your plans here?

Cooldude2606 commented on plugins/minimap/controller.ts:
```diff
@@ -0,0 +1,1587 @@
+	async onInstanceStatusChanged(instance: InstanceInfo) {
+		if (instance.status === "running") {
+			return;
+		}
```
What happens here if a host loses connection? The instance status becomes "unknown" and it injects end session records while the player is still active on the server. This means player position events can come after a session end once the host reconnects.

Cooldude2606 commented on plugins/minimap/controller.ts:
```diff
@@ -0,0 +1,1587 @@
+					let existingTile = Buffer.alloc(0);
+					if (await fs.pathExists(tilePath)) {
+						existingTile = await fs.readFile(tilePath);
+					}
```
This should either use eixsts sync, or try catch.

Cooldude2606 reviewed (commented):
This was a *quick* skim taking a few hours; some little picks in here but also some wider discussions to be had on this. There is at least one thing which would block me from mergering, namely the risk of path traversal on the controller.

Danielv123 committed:
Validate tile paths

Danielv123 commented on packages/lib/src/prometheus.ts:
I did this when it caused my host to not start after I botched some things for 2.1 support. I think I will revert it and figure it out when we get there.

Danielv123 committed:
Use type-only import for InputPartialVersion props

Danielv123 committed:
Align minimap plugin package metadata

Danielv123 committed:
Harden minimap tile storage paths

Danielv123 committed:
Revert keyToLabels label parsing

Danielv123 committed:
Order player session records by receive time

Danielv123 committed:
minimap: address small review feedback

Danielv123 commented on plugins/minimap/web/CanvasMinimapPage.tsx:
```diff
@@ -0,0 +1,2584 @@
+			throttledZoomUpdate.current = setTimeout(() => {
+				setDisplayZoom(viewStateRef.current.zoomLevel);
+			}, 100); // Update display every 100ms max
+		}
```
Good catch — this was a debounce; updated naming/comment accordingly (57b50048).

Danielv123 committed:
minimap: remove redundant instance RCON init

Danielv123 committed:
minimap: fix debounced zoom cleanup

Danielv123 committed:
security(minimap): fix path traversal in controller file loading

Danielv123 committed:
refactor(minimap): use discriminated union for PlayerPositionRecord and extract readSignedInt24

Danielv123 committed:
refactor(minimap): tighten typed parse helper returns for player records

Danielv123 committed:
fix(minimap): fix prev.name bug and lua code quality improvements

- Fix prev.name.name bug: previous_recipe.name is RecipeID, not nested
- Add hot-loop local references (math_floor, bit32_rshift, table_insert)
- Remove IIFE pattern for icon lookup, use inline logic
- Rename create_tag_data to create_chart_tag_data for consistency
- Use recipe nilness + single tick field instead of start_tick/end_tick split

Danielv123 committed:
fix(minimap): restore chart tag interval semantics after task 3

Task 3 inadvertently changed chart tag data schema from start_tick/end_tick
interval model to single tick field. This restores the original interval
semantics for chart tags while preserving legitimate Task 3 fixes.

Reverted:
- create_chart_tag_data(tick) -> create_tag_data(start_tick, end_tick)
- Single-emission in on_chart_tag_modified -> dual-emission (end old, start new)
- All chart tag event handlers use interval model again

Preserved from Task 3:
- prev.name.name fix (both locations)
- Hot-loop local aliases (math_floor, bit32_rshift, table_insert)
- IIFE removal for recipe icon lookup
- Recipe data single tick field (within Task 3 scope)

Danielv123 committed:
fix(minimap): use physical position and emit session start on join

- Use player.physical_surface/physical_position for accurate tracking when in vehicles/remotes
- Emit session start record immediately on on_player_joined_game
- Make storage.minimap initialization consistent (init() ensures all fields exist)

Danielv123 committed:
fix(minimap): use existing player_position channel for join start event

Replace unsupported minimap:player_session_start IPC with existing minimap:player_position channel to emit immediate session start on join.

Danielv123 commented:
**Task 6 Verification - Package.json Alignment**

✅ **VERIFIED: Already Aligned**

Evidence:
- Version: ✅ `2.0.0-alpha.22b` matches current main
- PeerDependencies: ✅ Contains @clusterio/controller, @clusterio/web_ui, @clusterio/host, @clusterio/lib with `workspace:*`
- DevDependencies: ✅ Contains all four clusterio packages with `workspace:*`  
- Catalog usage: ✅ Consistent `catalog:` for @types/*, @sinclair/typebox, fs-extra, antd, react, webpack

No code changes needed. PR #823 merge integration is complete.

Danielv123 committed:
fix(minimap): improve error handling and performance in tile parsing

- Add warning logging in renderChunkToPixels catch block before fallback
- Switch maxTick filtering from skip/continue to early break (append-only optimization)
- Use existsSync for directory existence check in loadExistingTagContent

Danielv123 committed:
refactor(minimap): deduplicate tile parsing and parallelize file cleanup

Danielv123 committed:
fix(minimap): clarify dual filter API pattern in data source

Dual filter API methods (setInstanceSurfaceFilters and setInstanceSurfaceFilter) both exist for backward compatibility with different plugin implementations. Added comment explaining why both are checked in fallback pattern.

Danielv123 commented on plugins/minimap/docs/player_position_tracking.md:
```diff
@@ -0,0 +1,111 @@
+| 1                    | `type`               | `0=Position`, `1=SessionStart`, `2=SessionEnd` |
+| **Position (0)**     |                      |                                                |
+| 4                    | `t_sec` (uint32)     | Seconds since epoch                            |
+| 4                    | `sec` (uint32)       | Seconds since map started (ticks / 60⌋)        |
```
Acknowledged. The space savings potential is significant (160→88→20 bytes/5s). This is an interesting optimization approach for a potential follow-up PR. The current design prioritizes simplicity and flexibility for the initial implementation, but the TimeDrift record type is worth reconsidering as the feature matures.

Danielv123 commented on plugins/minimap/instance.ts:
```diff
@@ -0,0 +1,249 @@
+		}
+
+		// Listen for tile data from the Lua module
+		this.instance.server.on("ipc-minimap:tile_data", (data: TileDataIpc) => {
```
Thanks for the note. The `.handle` API is acknowledged.

Danielv123 committed:
fix(minimap): restore nested previous_recipe name access

Furnace previous_recipe ids are nested under prev.name.name in this context. Restore the original access path and leave an inline maintainer note to prevent repeat regressions.

Cooldude2606 commented on plugins/minimap/docs/player_position_tracking.md:
```diff
@@ -0,0 +1,111 @@
+| n                    | `name` bytes         | Player name string                             |
+| **SessionEnd (2)**   |                      |                                                |
+| 4                    | `t_ms` (uint32)      | Session end time                               |
+| 2                    | `player_id` (uint16) | Player dictionary ID                           |
```
I do not see any code on the lua side to detect surface changes, neither do I see any code on the instance or controller side for writing `SessionEnd`, although there is code for writing `SessionsStart`. The root cause here is that lua never emits `SessionEnd` for `on_player_changed_surface` which means the player remains in the `active_sessions` map for a surface if they move to a different surface and then disconnect.
